### DISCUSSION TITLE
Websocket timeout

### DISCUSSION THREAD

**Discussion by fendixdev** [dani-garcia/vaultwarden#3881] [Q&A]
### Subject of the issue

With the latest update the websockets did not work anymore. I decided to reinstall, downloading the latest image and using the proxy template from BlackDex. When I check, I get errors in the console.

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.2
* Web-vault version: v2023.7.1
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.41.2
* Clients used: 1
* Reverse proxy and version: Nginx - v1.29.0+ (by @BlackDex), nginx version: nginx/1.18.0
* Other relevant information: -

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://**************",
  "domain_origin": "*****://**************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": false,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "******************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "**************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "******************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": "80365",
  "yubico_secret_key": "***",
  "yubico_server": null
}
```
</details>

### Steps to reproduce

Creating user and adding to docker group, create docker-compose.yml, create nginx vhost and obtain certificates with LetsEncrypt (certbot and dns-cloudflare challenge), nginx -t and restart, start container.

### Expected behaviour
LiveSync is not working.

### Actual behaviour
Websocket/LiveSync is not working. I'm receiving following errors in the browser console:

`[2023-09-12T08:27:52.672Z] Error: Connection disconnected with error 'Error: Server timeout elapsed without receiving a message from the server.'.`

`Unhandled Promise rejection: Server timeout elapsed without receiving a message from the server. ; Zone: <root> ; Task: null ; Value: Error: Server timeout elapsed without receiving a message from the server.
    at W.serverTimeout (HubConnection.ts:671:30)
    at HubConnection.ts:640:57
    at i.<computed> (zone.js:3158:47)
    at t.invokeTask (zone.js:446:35)
    at e.runTask (zone.js:214:51)
    at e.invokeTask (zone.js:528:38)
    at i.useG.invoke (zone.js:517:52)
    at r.args.<computed> (zone.js:3138:36) Error: Server timeout elapsed without receiving a message from the server.
    at W.serverTimeout (https://redacteddomain.de/app/vendor.80cb43550cd35d36022d.js:2:67226)
    at https://redacteddomain.de/app/vendor.80cb43550cd35d36022d.js:2:66870
    at i.<computed> (https://redacteddomain.de/app/polyfills.b65045dd8b47303eeac4.js:2:248497)
    at t.invokeTask (https://redacteddomain.de/app/polyfills.b65045dd8b47303eeac4.js:2:215383)
    at e.runTask (https://redacteddomain.de/app/polyfills.b65045dd8b47303eeac4.js:2:210412)
    at e.invokeTask (https://redacteddomain.de/app/polyfills.b65045dd8b47303eeac4.js:2:216539)
    at i.useG.invoke (https://redacteddomain.de/app/polyfills.b65045dd8b47303eeac4.js:2:216430)
    at r.args.<computed> (https://redacteddomain.de/app/polyfills.b65045dd8b47303eeac4.js:2:248195)`

`Error: Server timeout elapsed without receiving a message from the server.
    at W.serverTimeout (HubConnection.ts:671:30)
    at HubConnection.ts:640:57
    at i.<computed> (zone.js:3158:47)
    at t.invokeTask (zone.js:446:35)
    at e.runTask (zone.js:214:51)
    at e.invokeTask (zone.js:528:38)
    at i.useG.invoke (zone.js:517:52)
    at r.args.<computed> (zone.js:3138:36)`

### nginx vhost/reverse proxy
<details><summary>Show vhost</summary>

```bash
upstream vaultwarden-default {
  zone vaultwarden-default 64k;
  server 127.0.0.1:8080;
  keepalive 2;
}


map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      "";
}

server {
    listen 80;
    listen [::]:80;
    server_name redacteddomain.de;

    if ($host = redacteddomain.de) {
        return 301 https://$host$request_uri;
    }
    return 404;
}

server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;
    server_name redacteddomain.de;

    ssl_certificate /etc/letsencrypt/live/redacteddomain.de/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/redacteddomain.de/privkey.pem;
    ssl_trusted_certificate /etc/letsencrypt/live/redacteddomain.de/fullchain.pem;

    client_max_body_size 525M;

    location / {
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;

      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;

      proxy_pass http://vaultwarden-default;
    }
}
```
</details>

### docker compose file
<details><summary>Show docker-compose.yml</summary>

```bash
version: '3'

services:
  vaultwarden:
    image: vaultwarden/server:latest
    container_name: vaultwarden
    restart: always
    ports:
      - 8080:80
    environment:
      - ADMIN_TOKEN=******
      - SIGNUPS_ALLOWED=false
      - DOMAIN=https://redacteddomain.de
      - WEB_VAULT_ENABLED=true
      - ROCKET_PORT=80
      - YUBICO_CLIENT_ID=80365
      - YUBICO_SECRET_KEY=******
      - SMTP_HOST=smtp.gmail.com
      - SMTP_PORT=587
      - SMTP_USERNAME=******
      - SMTP_FROM=******
      - SMTP_PASSWORD=******
      - SMTP_SECURITY=starttls
    volumes:
      - ./vw-data:/data
```
</details>

Using cloudflare as a DNS service. No rules created. Strict encryption.

**Comment by BlackDex** [comment]
Are you also using CloudFlare as reverse proxy/entrypoint?
Because there are know issues with this for some reason.

**Comment by fendixdev** [comment]
Yes, all traffic is routed through cloudflare to hide the origin server.

**Reply by BlackDex** [reply]
I suggest to search the discussions for this, there are some people who have some remarks regarding this.

**Reply by fendixdev** [reply]
Thank you! I'm looking for tips in the meantime :)

Funny that it no longer works from one day to the next. But I also guess that Cloudflare makes problems there.

**Comment by valodzka** [comment]
Same issue, works without cloudflare but with cloudflare `Server timeout elapsed without receiving a message from the server`
I'm using websocket behind cloudflare successfully with other apps so it's something specific in vw/rust implementation that causes the issue.
Same described here #3835

**Reply by BlackDex** [reply]
@valodzka Could you maybe provide the headers sent/received from both Vaultwarden and the other app(s)?
Maybe there is something we can fix to resolve this issue, but i would need some data to compare.

**Reply by valodzka** [reply]
I tried to tcpdump connections and request headers (from browser/cloudflare) is the same. Response, from vw:

```
HTTP/1.1 101 Switching Protocols
sec-websocket-version: 13
sec-websocket-accept: mFx3L+aNUfedsL03eWdACrXWgCM=
server: Rocket
x-frame-options: SAMEORIGIN
permissions-policy: accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=(), geolocation=(), gyroscope=(), keyboard-map=(), magnetometer=(), microphone=(), midi=(), payment=(), picture-in-picture=(), screen-wake-lock=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=()
x-content-type-options: nosniff
referrer-policy: same-origin
x-xss-protection: 0
content-security-policy: default-src 'self'; base-uri 'self'; form-action 'self'; object-src 'self' blob:; script-src 'self' 'wasm-unsafe-eval'; style-src 'self' 'unsafe-inline'; child-src 'self' https://*.duosecurity.com https://*.duofederal.com; frame-src 'self' https://*.duosecurity.com https://*.duofederal.com; frame-ancestors 'self' chrome-extension://nngceckbapebfimnlniiiahkandclblb chrome-extension://jbkfoedolllekgbhcbcoahefnbanhhlh moz-extension://* ; img-src 'self' data: https://haveibeenpwned.com https://www.gravatar.com ; connect-src 'self' https://api.pwnedpasswords.com https://api.2fa.directory https://app.simplelogin.io/api/ https://app.anonaddy.com/api/ https://api.fastmail.com/ https://api.forwardemail.net ;
cache-control: no-cache, no-store, max-age=0
connection: Upgrade
upgrade: websocket
content-length: 0
date: Mon, 18 Sep 2023 10:44:33 GMT
```

Applications that works, one written in Go:

```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: t6n7bG7l5Qgof/L7Na/VLIP2mQY=
```

One written in Ruby:

```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: WqKCyJlEYh4H2BxOQ2JzSVEjoxU=
```

**Reply by BlackDex** [reply]
Hmm, maybe the headers sent like `x-xss-protection`, `referrer-policy` etc.. etc.. are causing issues.
Thanks! This was really helpful.

**Reply by valodzka** [reply]
I tried add this to nginx (for `location /vw/notifications/hub`) and it works now, so yes, it's headers issue:

  ```
       proxy_hide_header sec-websocket-version;
        proxy_hide_header server;
        proxy_hide_header x-frame-options;
        proxy_hide_header permissions-policy;
        proxy_hide_header x-content-type-options;
        proxy_hide_header referrer-policy;
        proxy_hide_header x-xss-protection;
        proxy_hide_header content-security-policy;
        proxy_hide_header cache-control;
        proxy_hide_header content-length;
        proxy_hide_header date;
```

**Reply by BlackDex** [reply]
Thanks! I think this would be an easy fix.

**Reply by ccben87** [reply]
WOW. I just came across the strangest issue where my requests were timing out going from Kubernetes pod in one cluster to another Kubernetes pod in a different cluster but was resolved by hiding the headers as suggested by @valodzka (thanks so much btw!). It works going from K8s pod to K8s pod in the same cluster.  As of typing this I hid them for / and not /vw/notifications/hub because I was never able to load Vaultwarden at all.

No issue when loading from a standard Linux system or between two pods in the one K8s cluster but when loading from one K8s pod in one cluster to another K8s pod in a separate cluster, it would time out. I traced the connection from one pod in one K8s cluster going all the way back to Cilium which logs say delivered it to the destination pod yet my Curl or webserver in that K8s cluster would just time out and act as if it never received the reply. This issue has been driving me up the wall for days since, otherwise, anything else that was running as a test worked just fine.

I have no clue about what those headers actually mean or which header. if not all of them, which results in these bizarre timeouts but I do intend to troubleshoot further to narrow down what is happening here. 

@BlackDex are these headers required to use Websocket or not? Can we safely drop them?

**Reply by BlackDex** [reply]
In testing this is already fixed (i think).
Maybe @valodzka can test this by running testing and remove all those `proxy_hide_headers`

**Reply by ccben87** [reply]
> In testing this is already fixed (i think). Maybe @valodzka can test this by running testing and remove all those `proxy_hide_headers`

Oh, that's good. @BlackDex How stable would you say testing is at the moment since I'd rather hide the headers and have a more stable experience as I'm about to put a system into production. I'm happy to test either way.

**Reply by BlackDex** [reply]
And you are going through Cloudflare?
If so please test if the `testing` tagged images work without any modifications, since we drop/exclude those headers for the websocket connections.

I think it is stable, but maybe i'm not a good reference point haha.
There are probably some quirks still, but there are quirks on stable too which are fixed in testing.
So, it's a bit up to you i think.

There might be some issues we do not know about yet. But those will only appear if someone is going to use it.

**Reply by ccben87** [reply]
> And you are going through Cloudflare?

Nope. Not Cloudflare here. Going from one K8s cluster (Talos Linux with Cilium CNI) with nodes running in VMs to another K8s cluster (Talos Linux with Cilium CNI) running in VMs, all VMs running on the same physical Nutanix AHV cluster.

**Reply by ccben87** [reply]
I'll give your docker "testing-alpine" a go.

**Reply by ccben87** [reply]
@BlackDex 

Okay, I've tried testing-alpine and I have to use the following in nginx to even get a response back from vaultwarden:

        location / {
            proxy_pass http://<snip>:8080 ;
            proxy_hide_header x-content-type-options;
            proxy_hide_header referrer-policy;
            proxy_hide_header x-frame-options;
            proxy_hide_header permissions-policy;
            proxy_hide_header content-security-policy;
            proxy_hide_header cache-control;
            proxy_hide_header x-xss-protection;


        }



The only problem is, the same as on stable version, is that the first page doesn't actually load in a web browser but it works in curl. Not sure why. Maybe if I adjust the above to only be for a certain path. 

My browser just sits there not ever loading the following:
```
https://<snip>/app/main.f9a460fbbcf515d25766.css
https://<snip>/app/polyfills.3409b07d989eea690b2c.js
https://<snip>/app/vendor.248a2525a0f5c30c5651.js
https://<snip>/app/main.d4fab1a39a45b0dc4150.js
```

Any advice on my situation would be much appreciated. I've really struggled with this issue.

**Reply by BlackDex** [reply]
Could be cache in this case, since the newer images uses a newer web-vault, and there the names are different.

**Reply by ccben87** [reply]
If you've got any suggestions about anything I can test, I'd much appreciate it.

I note that with all the hidden headers above that Vaultwarden works directly but not when going K8s to K8s even after hiding the headers. I do get a response back when the headers are hidden but it just never seems to load the above JS files. nginx logs do say:

`*202 an upstream response is buffered to a temporary file /tmp/proxy_temp/0000000041 while reading upstream, client
`

For context, we have a reverse proxy (Kong) that runs in K8s in one cluster and Vaultwarden in a separate cluster behind nginx. We've tried removing the extra proxy layer but that makes no difference.

**Reply by BlackDex** [reply]
But do you need to remove those from all requests?
Not only websocket requests? Because those headers are there for security

**Reply by ccben87** [reply]
> But do you need to remove those from all requests? Not only websocket requests? Because those headers are there for security

I'm not entirely sure but I note applying it to just "/vw/notifications/hub" doesn't help me at all. If I set it to "/" then at least I get a reply rather than a timeout but then the page never loads even after applying it to "/" anyway. Not really sure what's going on. I'd love to have those features turned on, for security reasons, yeah. At least in our case we will be putting it behind a WAF but still.

**Reply by BlackDex** [reply]
Please keep in mind that WAF's do tend to add/override headers sometimes which break the functionality of Vaultwarden/Bitwarden clients.
That might be the case here already.

Best would be to check the Browsers Developer Console and see what the issues are.

**Reply by ccben87** [reply]
> Please keep in mind that WAF's do tend to add/override headers sometimes which break the functionality of Vaultwarden/Bitwarden clients. That might be the case here already.

For the sake of testing I'm just using pure nginx to nginx to eliminate that as a factor.

In my browser I get 

```
60075ms- 504 gateway timeout - http://<snip>/app/polyfills.3409b07d989eea690b2c.js
0ms response time - 200 -  http://<snip>/theme_head.1df11f603fda400762b7.js
6000ms response time - 200 - http://<snip>/app/main.f9a460fbbcf515d25766.css
6000ms response time - 200 - http://<snip>/app/vendor.248a2525a0f5c30c5651.js
60075ms- 504 gateway timeout - http://<snip>/app/main.d4fab1a39a45b0dc4150.js
117117ms - 200 - http://<snip>/images/apple-touch-icon.png
0 ms - 200 - http://<snip>/images/favicon-16x16.png
```

The displayed page text just says "Vaultwarden logo"



```
GEThttp://<snip>/app/polyfills.3409b07d989eea690b2c.js
[HTTP/1.1 504 Gateway Time-out 60075ms]

GEThttp://<snip>/app/main.d4fab1a39a45b0dc4150.js
[HTTP/1.1 504 Gateway Time-out 60117ms]

Layout was forced before the page was fully loaded. If stylesheets are not yet loaded this may cause a flash of unstyled content. [node.js:393](resource://devtools/server/actors/inspector/node.js)
Loading failed for the <script> with source “http://<snip>/app/polyfills.3409b07d989eea690b2c.js”. [<snip>:1:1](<snip>/)
Loading failed for the <script> with source “http://<snip>/app/vendor.248a2525a0f5c30c5651.js”. [<snip>:1:1](<snip>/)
Loading failed for the <script> with source “http://<snip>/app/main.d4fab1a39a45b0dc4150.js”. [<snip>:1:1](<snip>/)
Source map error: Error: request failed with status 504
Resource URL: http://<snip>/theme_head.1df11f603fda400762b7.js
Source Map URL: theme_head.1df11f603fda400762b7.js.map
```

**Reply by BlackDex** [reply]
Looks more like an issue with either your network or ports or firewall or at least something not being able to reach the Vaultwarden container in some way.

**Reply by ccben87** [reply]
Unfortunately I've traced it all the way back to the pod from packet capture on the firewall and monitoring on the Cilium CNI where I can see that it is being delivered to the pod according to its log. Traffic is definitely flowing both ways. I wasn't able to get any response at all until I hid those headers and now I get some response but others just time out. Those 200s for the scripts must be incorrect btw because the size I got back is 0b, must be something that happened from the timeout and connection being closed.

**Reply by valodzka** [reply]
@BlackDex I did quick test for `testing` image, results:
- live updates work with web app
- live updates doesn't work in chrome extension

With proxy_hide_headers and old version they work in both cases.

**Reply by ccben87** [reply]
Okay so turns out it was an issue with the MTU. All is working with default headers so far but I haven't tested what @valodzka has so I assume what he says isn't working is also not working in my case. I'm going to continue using the testing docker image
